gomodcmd.go

```diff
@@ -56,29 +58,39 @@ func listModules(modules ...string) (mods map[string]*listModule, err error) {
 	return mods, nil
 }
 
-func goModInfo() (*modfile.File, error) {
-	// Get info on the current module so that we can find the
-	// path to the current go.mod file.
-	mods, err := listModules()
-	if err != nil {
-		return nil, errors.Wrap(err)
-	}
-	if len(mods) != 1 {
-		return nil, errors.Newf("unexpected module list count (want 1 got %d)", len(mods))
-	}
-	var goModPath string
-	for _, m := range mods {
-		goModPath = m.GoMod
+// goModInfo returns the main module's root directory
+// and the parsed contents of the go.mod file.
+func goModInfo() (string, *modfile.File, error) {
+	rootDir := findModuleRoot(cwd)
+	if rootDir == "" {
+		return "", nil, errors.Newf("no main module found")
 	}
+	goModPath := filepath.Join(rootDir, "go.mod")
 	data, err := ioutil.ReadFile(goModPath)
 	if err != nil {
-		return nil, errors.Notef(err, nil, "cannot read current go.mod file")
+		return "", nil, errors.Notef(err, nil, "cannot read main go.mod file")
 	}
 	modf, err := modfile.Parse(goModPath, data, nil)
 	if err != nil {
-		return nil, errors.Wrap(err)
+		return "", nil, errors.Wrap(err)
+	}
+	return rootDir, modf, nil
+}
+
+func findModuleRoot(dir string) (root string) {
+	dir = filepath.Clean(dir)
+
+	// Look for enclosing go.mod.
+	for {
+		if fi, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil && !fi.IsDir() {
```

Comment:
why not check `fi.IsRegular`? that seems much safer to me. I don't know if symlinks would work, and you definitely don't want to worry about other file types like named pipes.

Reply:
This is exactly the same code as is currently in the `go` command. I think it's probably good to keep it the same so the results are sure to be identical.